### PR TITLE
Meter gap

### LINKED ISSUE #7522: Meter -- support "gap" between items
To support accessibility, Meter should provide ability for a "gap" to be added in between values.

Desired functionality:

Image

Image

d3 charts has the desired capability: https://d3js.org/d3-shape/arc#arc_padAngle

### PR DESCRIPTION
**Pull request by MikeKingdom**
#### What does this PR do?
This PR adds a gap between segments on the various meter styles. This gap is intended to increase visible differentiation between the colored segments. The width of the gap is specified in the theme settings for Meter and currently defaults to `xxsmall` which will be 3pixels generally.

_Note: we may consider having the base theme have no gap and only put the gap in HPE theme versions if we want to limit impact, but it seems like this would be good for accessibility for all consumers._

The gap only comes into play when there are multiple segments in the meter. 

#### Where should the reviewer start?
components/Meter/Circle.js

#### What testing has been done on this PR?
Storybook as well as custom examples to test edge cases.

#### How should this be manually tested?
Storybook Visualization/Meter examples (Pie and Vertical Bar Multiple especially)
image
image
image
image
image
image
image

#### Do Jest tests follow these best practices?

- [ ] `screen` is used for querying.
- [ ] The correct query is used. (Refer to [this list of queries](https://testing-library.com/docs/queries/about/#priority))
- [ ] `asFragment()` is used for snapshot testing.

#### Any background context you want to provide?
The circular meters draw the gap as a constant width separate between the segments. This can make segments smaller than about 4 or so pixels less visible since the separator takes some space.

#### What are the relevant issues?
Fixes #7522 

#### Do the grommet docs need to be updated?
Yes (Meter theme now includes `gap` value)

#### Should this PR be mentioned in the release notes?
Yes

#### Is this change backwards compatible or is it a breaking change?
Backwards compatible

### CODE REVIEW

File: src/js/components/Meter/Bar.js [hunk 1]
```diff
@@ -24,6 +24,10 @@ const Bar = forwardRef((props, ref) => {
   const thickness = parseMetricToNum(
     theme.global.edgeSize[thicknessProp] || thicknessProp,
   );
+
+  const gapTheme = theme.meter?.gap ?? 'xxsmall';
```
Comment (taysea) [inline src/js/components/Meter/Bar.js]: Could we define the "xxsmall" as the default value in base.js? Then just have:

```suggestion
  const gapTheme = theme.meter?.gap;
```

File: src/js/components/Meter/Circle.js [hunk 2]
```diff
@@ -30,13 +34,18 @@ const Circle = forwardRef((props, ref) => {
   //  (type === 'semicircle' ? 180 : 360) / max;
   const someHighlight = (values || []).some((v) => v.highlight);
 
+  const gapTheme = theme.meter?.gap ?? 'xxsmall';
```
Comment (taysea) [inline src/js/components/Meter/Circle.js]: Same comment as above.

File: src/js/components/Meter/Bar.js (hunk 1 above)
Comment (taysea) [reply src/js/components/Meter/Bar.js]: Or I suppose I see it is already defined in base.js, so are we able to just remove the `??`?

File: src/js/themes/base.d.ts [hunk 3]
```diff
@@ -1274,6 +1274,7 @@ export interface ThemeType {
     color?: ColorType;
     colors?: GraphColorsType;
     extend?: ExtendType;
+    gap?: string;
```
Comment (taysea) [inline src/js/themes/base.d.ts]: Should we use this?

```suggestion
    gap?: GapType;
```

**Review by taysea — COMMENTED** [review]
The direction this is heading looks great!

File: src/js/components/Meter/Bar.js (hunk 1 above)
Comment (MikeKingdom) [reply src/js/components/Meter/Bar.js]: I replaced this with
```jsx
const gapTheme = theme.meter?.gap ?? 0;
```
This just keeps it from crashing if meter.gap somehow didn't end up in the theme. I'm not sure how likely that really is but this makes it not a problem either way.

File: src/js/components/Meter/Circle.js (hunk 2 above)
Comment (MikeKingdom) [reply src/js/components/Meter/Circle.js]: same

File: src/js/themes/base.d.ts (hunk 3 above)
Comment (MikeKingdom) [reply src/js/themes/base.d.ts]: Done

**Comment by MikeKingdom** [comment]
> The direction this is heading looks great!
> 
> EDIT: Based on offline discussion, would prefer if possible an approach that doesn't overlay the gaps but instead draws the segments in a way that leaves a gap between.

Same :) Refactored to just draw the segments appropriately leaving gaps and not try to overlay a separate gap.

File: src/js/themes/base.js [hunk 4]
```diff
@@ -1282,6 +1282,7 @@ export const generate = (baseSpacing = 24, scale = 6) => {
       color: 'graph-0',
       // colors: [] || colors: ['graph-0', 'graph-1', 'graph-2', 'graph-3'],
       // extend: undefined,
+      gap: 'xxsmall',
```
Comment (taysea) [inline src/js/themes/base.js]: Wondering (for backwards compatibility) if we shouldn't touch base.js and just turn this on in grommet-theme-hpe?

Comment (MikeKingdom) [reply src/js/themes/base.js]: I was debating that too (note in the original description.)

**Comment by taysea** [comment]
In the Vertical Bar with multiple values, if I turn on `round`, for some reason the value paths are shifted upwards from the edge (changed background to `red` just to highlight what I'm talking about) -- This only happens when `direction="vertical"`

Red background for visual emphasis
Screenshot 2025-07-24 at 3 41 30 PM

Original background color
Screenshot 2025-07-24 at 3 41 00 PM

File: src/js/components/Meter/stories/Circle.stories.js [hunk 5]
```diff
@@ -25,7 +19,9 @@ export const Circle = () => {
       <Meter
         type="circle"
         background="light-2"
-        values={[{ value, color: value > 50 ? 'accent-2' : 'accent-1' }]}
+        values={[
+          { value, color: value > 50 ? 'status-critical' : 'status-ok' },
```
Comment (taysea) [inline src/js/components/Meter/stories/Circle.stories.js:23]: Locally I increase `theme.meter.gap` to "medium" and for some reason when the value first reaches greater than 50, the path is offset.

It's mostly apparent when `round={true}` but also happens subtly on the non-round version too.

Screenshot 2025-07-24 at 3 50 11 PM

Comment (taysea) [reply src/js/components/Meter/stories/Circle.stories.js:23]: The other thing that's interesting/unexpected it the rounding at the start of the value path. It's less rounded than the end and I think I would expect it to be fully rounded

File: src/js/components/Meter/utils.js [hunk 6]
```diff
@@ -19,6 +19,25 @@ export const strokeProps = (color, theme) => {
   return result;
 };
 
+export const fillProps = (color, theme) => {
+  const result = {};
+  if (color) {
+    if (typeof color === 'object') {
+      result.fill = normalizeColor(color.color, theme);
+      // if (color.opacity) {
```
Comment (taysea) [inline src/js/components/Meter/utils.js]: Is this needed? or remove

File: src/js/components/Meter/stories/Circle.stories.js (hunk 5 above)
Comment (MikeKingdom) [reply src/js/components/Meter/stories/Circle.stories.js:23]: That is the gap shape basically and it's more exaggerated when it's bigger (ie. that edge won't point to the center). I probably should avoid any of the gap where it's against the track rather than another segment

Comment (MikeKingdom) [reply src/js/components/Meter/stories/Circle.stories.js:23]: > The other thing that's interesting/unexpected it the rounding at the start of the value path. It's less rounded than the end and I think I would expect it to be fully rounded

Yup. That's a bug. For all but the first segment, that rounded end goes the other way and is less rounded so it forms a gap with the previous segment. For the first segment it's supposed to be the same roundness as the end.

File: src/js/components/Meter/utils.js (hunk 6 above)
Comment (MikeKingdom) [reply src/js/components/Meter/utils.js]: So the stroke seemed to support an opacity from the theme. Now that's it's a filled shape I'd think we'd need to still support that opacity now for the fill?

Comment (taysea) [reply src/js/components/Meter/utils.js]: Yes agree to continuing to support opacity

File: src/js/components/Meter/stories/Circle.stories.js (hunk 5 above)
Comment (MikeKingdom) [reply src/js/components/Meter/stories/Circle.stories.js:23]: I improved this a lot and made it smarter, especially for single segment meters. A 50% segment will look much better. With multi-segment meters that are rounded, you can still get some overlap that is interesting. It pretty much mimics existing behavior.

**Comment by taysea** [comment]
all looks great -- one interesting thing is this case, might expect the rounding of the first segment to be inverse:

but I want to be mindful of not sinking too much time into this -- open if you think it's worthwhile to handle now or improve in future.

Screenshot 2025-07-28 at 8 59 44 AM

**Comment by MikeKingdom** [comment]
> all looks great -- one interesting thing is this case, might expect the rounding of the first segment to be inverse:
> 
> but I want to be mindful of not sinking too much time into this -- open if you think it's worthwhile to handle now or improve in future.

I kept the start rounding of the first segment as rounded outward since that is how it is before this gap pr. You can see the effect in COM's dashboard as a live example. I think the rounding thing is generally weird in a circular meter, but that's what the prop is for :) I'm assuming we should keep that as is given its history and pre-existing usage.

Looks like I may still have the wrong radius on the start of the first segment (only obvious as the gap is larger).

File: src/js/themes/base.js (hunk 4 above)
Comment (britt6612) [reply src/js/themes/base.js]: I wonder though would this effect backwards compatibility if were "fixing" to account for accessibility?

**Comment by britt6612** [comment]
Screenshot 2025-07-31 at 1 05 39 PM


you might have covered this (little late to some of the discussion) but did we chat through a implementation as above.

```
      <Meter type="circle" values={values} round thickness="medium" />

```
I believe this is happening because the round extends the stroke beyond the arc path?

**Comment by MikeKingdom** [comment]
> you might have covered this (little late to some of the discussion) but did we chat through a implementation as above.
> 
> ```
>       <Meter type="circle" values={values} round thickness="medium" />
> ```
> 
> I believe this is happening because the round extends the stroke beyond the arc path?

It's basically keeping with existing functionality/style. The first segment starts with a rounded outward cap and if the meter is full it overlaps like that. It's the original style. COM has meters like this. Personally I don't like rounded circle meters, but they exist.

**Comment by britt6612** [comment]
> > you might have covered this (little late to some of the discussion) but did we chat through a implementation as above.
> > ```
> >       <Meter type="circle" values={values} round thickness="medium" />
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I believe this is happening because the round extends the stroke beyond the arc path?
> 
> It's basically keeping with existing functionality/style. The first segment starts with a rounded outward cap and if the meter is full it overlaps like that. It's the original style. COM has meters like this. Personally I don't like rounded circle meters, but they exist.

gotcha thanks for clarifying since its the original style I agree with leaving just noticed so thought Id ask

**Comment by britt6612** [comment]
This is very detailed but it seems as though the `gaps` are a little inconsistent. see screen shot. Im wondering if this is how the startGap and endGap is calculated?

Screenshot 2025-07-31 at 3 15 10 PM

**Comment by MikeKingdom** [comment]
> This is very detailed but it seems as though the `gaps` are a little inconsistent. see screen shot. Im wondering if this is how the startGap and endGap is calculated?

Any hints how you got that to happen? It looks like it's probably a `type="pie"`. What are the `values` used and `size` and `max`? What `gap` in the theme?

Long answer: This would be easier with pictures but... :)

Basically the 3 (or 4 for `type="circle"`) points that make up each slice are polar coordinates: an angle and radius from the center (that's the same as the original code.) But to add the gap we calculate the angle to add or substract from that angle such that the distance from that point is the value of `gap` from that original point (see `calcAngle()`. Once we have those points we can covert them from polar coords to x,y coords (see `polarToCartesian()`) and draw a couple lines for the start and end sides and a couple arcs connecting the sides.

For `type=circle` (aka annular ring) we have to do that twice, once for the outer arc and once for the inner arc.

For `type="pie"` those inner endpoints are basically an angle of +/- 90 degrees from the angle and a radius of `gap` from the center. The start and end lines through those points end up crossing near-ish the center if the slice is < 180 degrees. In that case we calculate the `intersection()` of those points since that will be where we actually want the point of the slice.

The result should be adjacent slices sides are parallel to each other since they are based off the same angle at that spot. 

My attempt to reproduce your example ended up like this:
image

**Comment by MikeKingdom** [comment]
> This is very detailed but it seems as though the `gaps` are a little inconsistent. see screen shot. Im wondering if this is how the startGap and endGap is calculated?

Turns out this example has a couple background color slices (values) so they look wonky/irregular next to real gaps or colored slices. Really not a valid use case.

File: src/js/components/Meter/Bar.js [hunk 7]
```diff
@@ -24,27 +24,66 @@ const Bar = forwardRef((props, ref) => {
...
         const key = `p-${index}`;
-        const delta = (value * (length - 2 * capOffset)) / max;
-        const d =
+        const delta = (value * lengthAvailable) / max;
+
+        // add a little bit extra to start to allow for larger rounded inset cap
+        // The extra needed can be calculated by the Pythagorean theorem
+        const extraGap =
+          round && index !== 0
+            ? Math.sqrt((thickness / 2 + gap / 4) ** 2 - (thickness / 2) ** 2)
```
Comment (britt6612) [inline src/js/components/Meter/Bar.js:54]: I might be overlooking but why divide by 4?

Comment (MikeKingdom) [reply src/js/components/Meter/Bar.js:54]: It's basically how the math works out for using a larger radius on the start of the next segment's inwardly rounded cap

Comment (britt6612) [reply src/js/components/Meter/Bar.js:54]: gotcha thanks!

**Review by britt6612 — APPROVED** [review]
LGTM

**Review by jcfilben — APPROVED** [review]
Looks good!

**Comment by Sulaymon333** [comment]
I know this is closed and merged. But while doing QA testing for next Grommet version release, I noticed that once a very large gap is passed, it looks like some segments are gone or almost gone. The question is, should we limit a range of gap tshirt sizes like from [xxsmall ... medium] for example to only be the acceptable values?

image
image

image
image

**Comment by MikeKingdom** [comment]
Thanks Sulaymon. We don't tend to put limits on dimensions like this, especially from the theme. This is one of those situations where someone overriding this value will figure out what works for them.
